Ensure that on first call to ReadService.stateUpdates beginOffset is …

simonmaxen-da opened the pull request:
…None

Fixes #6023

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

SamirTalwar-DA commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala:
```diff
@@ -110,11 +110,15 @@ private class JdbcLedgerDao(
-        .as(offset("ledger_end").single)
+        .as(offset("ledger_end").?.map(_.getOrElse(Offset.begin)).single)
     }
+  }
```
The braces are unnecessary.

simonmaxen-da replied:
Fixed.

SamirTalwar-DA commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala:
```diff
@@ -110,11 +110,15 @@ private class JdbcLedgerDao(
     dbDispatcher.executeSql(metrics.daml.index.db.getLedgerEnd) { implicit conn =>
       SQL_SELECT_LEDGER_END
-        .as(offset("ledger_end").single)
+        .as(offset("ledger_end").?.map(_.getOrElse(Offset.begin)).single)
```
I think that when this function is called, there should _always_ be a value here, so crashing would be appropriate. If we haven't initialized the ledger, the code should be calling `lookupInitialLedgerEnd` instead.

What happens if you revert this? What breaks?

gerolf-da replied:
Yes, that's correct, @SamirTalwar-DA.

simonmaxen-da replied:
The previous behaviour was:
1. If called prior to ledger initialisation - _crash_
2. If called post initialisation but pre first ledger end update - _default to Offset.begin_
3. If called post first ledger end update - _return most recently updated ledger_end_

This change means that the default provided in #2 above comes from the method as opposed to from the database column.

As many things could be happening asynchronously during startup I think it is safest to leave the default, or leave it someone more expert in the area to remove it ;-)

SamirTalwar-DA commented:
Fixes #6023.

simonmaxen-da commented:
@SamirTalwar-DA - I've updated the comment wrt 'Fixes'.  More importantly I've reverted some poor advice I previously gave in CONTRIBUTING.md on this subject.

SamirTalwar-DA commented:
You need to update the PR description, not the commit message. 😉

SamirTalwar-DA commented on CONTRIBUTING.md:
```diff
@@ -21,7 +21,7 @@ For Git commit messages, our principle is that `git log --pretty=oneline` should
   * *Remove module X because it is not used.*
 * Commits should have a description that concisely explains the rationale and context for the change if that is not obvious.
-* The first line of the description should include a reference to the fixed issue `(#XXX)`.
+* Commit descriptions should include a `Fixes #XX` line indicating what GitHub issue number the commit fixes.
```
```suggestion
* Commit descriptions can include a `Fixes #XX` line indicating what GitHub issue number the commit fixes.
```

simonmaxen-da replied:
As discussed with Samir this was a revert of CONTRIBUTING.md.

SamirTalwar-DA commented on CONTRIBUTING.md:
```diff
@@ -35,14 +35,14 @@ For Git commit messages, our principle is that `git log --pretty=oneline` should
 The following is an example of a well-formed commit, including the description (first line) and a body that includes changelog additions:
 
-      Introduced a new API for package management (#1311)
+      Fixes #1311
```
That's not a useful first line. The previous one was better.

We tend to use the present tense in commit messages, so I'd suggest "Introduces" or "Introduce" over "Introduced".

```suggestion
      Introduces a new API for package management

      Fixes #1311.
```

simonmaxen-da replied:
as above

SamirTalwar-DA commented on CONTRIBUTING.md:
```diff
@@ -35,14 +35,14 @@ For Git commit messages, our principle is that `git log --pretty=oneline` should
-      Introduced a new API for package management (#1311)
+      Fixes #1311
 
-      Also fixes a typo in the Scala bindings documentation
+      Also fixes a typo in the Scala bindings documentation.
 
       CHANGELOG_BEGIN
 
       - [Sandbox] Introduced a new API for package management.
-      See `#1311 <https://github.com/digital-asset/daml/issues/1311>`__.
+        See `#1311 <https://github.com/digital-asset/daml/issues/1311>`__.
 
       CHANGELOG_END
```
I have no idea why this document uses 6 spaces for indenting code, but it really confuses GitHub. I suggest changing it to use 4 spaces.

simonmaxen-da replied:
as above

SamirTalwar-DA commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala:
```diff
@@ -139,7 +141,7 @@ private class JdbcLedgerDao(
   // is advancing it.
   private val SQL_UPDATE_LEDGER_END = SQL(
-    "update parameters set ledger_end = {LedgerEnd} where ledger_end < {LedgerEnd}")
+    "update parameters set ledger_end = {LedgerEnd} where (ledger_end is null or ledger_end < {LedgerEnd})")
```
```suggestion
    "update parameters set ledger_end = {LedgerEnd} where ledger_end is null or ledger_end < {LedgerEnd}")
```

simonmaxen-da replied:
as above
